utf8.c

```diff
@@ -1726,25 +1773,24 @@ Perl__utf8n_to_uvchr_msgs_helper(const U8 *s,
                     ||  (flags & (UTF8_WARN_SUPER|UTF8_WARN_PERL_EXTENDED)))
                 {
 
-                    /* The warnings code explicitly says it doesn't handle the
-                     * case of packWARN2 and two categories which have
-                     * parent-child relationship.  Even if it works now to
-                     * raise the warning if either is enabled, it wouldn't
-                     * necessarily do so in the future.  We output (only) the
-                     * most dire warning */
-                    if (! (flags & UTF8_CHECK_ONLY)) {
-                        if (msgs || ckWARN_d(WARN_UTF8)) {
-                            pack_warn = packWARN(WARN_UTF8);
-                        }
-                        else if (msgs || ckWARN_d(WARN_NON_UNICODE)) {
-                            pack_warn = packWARN(WARN_NON_UNICODE);
-                        }
-                        if (pack_warn) {
-                            message = Perl_form(aTHX_ "%s: %s (overflows)",
+                    /* Overflow is a hybrid. If the word size on this platform
+                     * were wide enough for this to not overflow, a non-Unicode
+                     * code point would have been generated.  If the caller
+                     * wanted warnings for such code points, the warning
+                     * category would be WARN_NON_UNICODE, On the other hand,
+                     * overflow is considered a malformation, which is serious,
+                     * and the category would be just WARN_UTF8.  We clearly
+                     * should warn if either category is enabled, but which
+                     * category to use?  Historically, we've used 'utf8' if it
+                     * is enabled; and that seems like the more severe
+                     * category, more befitting a malformation. */
+                    pack_warn = NEED_MESSAGE(WARN_UTF8,
+                                             ckWARN_d, WARN_NON_UNICODE);
```

Comment:
This is not the same logic. In the old version, if `msgs` was true, `ckWARN_d(WARN_UTF8)` false, `ckWARN_d(WARN_NON_UNICODE)` true, we still used `WARN_UTF8`. With the new macro, we now use `WARN_NON_UNICODE`.